cmd/pod-scaler/resources.go

```diff
@@ -0,0 +1,121 @@
+package main
+
+import (
+	"context"
+	"sync"
+
+	"github.com/openhistogram/circonusllhist"
+	"github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/test-infra/prow/interrupts"
+	"k8s.io/test-infra/prow/pjutil"
+
+	pod_scaler "github.com/openshift/ci-tools/pkg/pod-scaler"
+)
+
+func newResourceServer(cpu, memory []*cacheReloader, health *pjutil.Health) *resourceServer {
+	logger := logrus.WithField("component", "request_server")
+	server := &resourceServer{
+		logger:     logger,
+		lock:       sync.RWMutex{},
+		byMetaData: map[pod_scaler.FullMetadata]corev1.ResourceRequirements{},
+	}
+	var infos []digestInfo
+	for i := range cpu {
+		infos = append(infos, digestInfo{name: cpu[i].name, data: cpu[i], digest: server.digestCPU})
+	}
+	for i := range memory {
+		infos = append(infos, digestInfo{name: memory[i].name, data: memory[i], digest: server.digestMemory})
+	}
+	loadDone := digest(logger, infos...)
+	interrupts.Run(func(ctx context.Context) {
+		select {
+		case <-ctx.Done():
+			logger.Debug("Waiting for readiness cancelled.")
+			return
+		case <-loadDone:
+			logger.Debug("Ready to serve resource request recommendations.")
+			health.ServeReady()
+		}
+	})
+
+	return server
+}
+
+type resourceServer struct {
+	logger *logrus.Entry
+	lock   sync.RWMutex
+	// byMetaData caches resource requirements calculated for the full assortment of
+	// metadata labels.
+	byMetaData map[pod_scaler.FullMetadata]corev1.ResourceRequirements
+}
+
+const (
+	// cpuRequestQuantile is the quantile of CPU core usage data to use as the CPU request
+	cpuRequestQuantile = 0.8
+)
+
+func formatCPU() toQuantity {
+	return func(valueAtQuantile float64) *resource.Quantity {
+		return resource.NewMilliQuantity(int64(valueAtQuantile*1000), resource.DecimalSI)
+	}
+}
+
+func (s *resourceServer) digestCPU(data *pod_scaler.CachedQuery) {
+	s.logger.Debugf("Digesting new CPU consumption metrics.")
+	s.digestData(data, cpuRequestQuantile, corev1.ResourceCPU, formatCPU())
+}
+
+const (
+	// memRequestQuantile is the quantile of memory usage data to use as the memory request
+	memRequestQuantile = 0.8
```

Comment:
Super pedantic, but do we need separate consts for cpu/mem since they are both `0.8`? Is the thought that they will diverge at some point?

Reply:
I do expect us to change these moving forward, and perhaps diverge them - we may be interested in making our CPU overcommit smaller (since it's much more fungible), etc.